File: judge/python/tests/bst_to_dlist.py
from common import *
from solution import *
import copy
import sys
import datetime

num_test = 100
true, false = True, False
in_0 = []
in_org_0 = []
out = []

def check_dlist(arr, list):
	cur, pre, i = list, None, 0

	while(cur != None):
		if (i >= len(arr) or arr[i] != cur.val): return False
		i += 1
		pre = cur
		cur = cur.right

	if i != len(arr): return False
	cur = pre
	i -= 1
	while(cur != None):
		if (i < 0 or arr[i] != cur.val): return False
		i -= 1
		last = cur
		cur = cur.left
		last.left = None

	if i != -1: return False
	return True

def dlist_to_string(list):
	ret = "["
	first = True
	while(list != None):
		if not first: ret += ", "
		ret += str(list.val)
		list = list.right
		first = False
	return ret + "]"

def load_test():
    f = open('judge/tests/bst-to-dlist.txt', 'r')
    global in_0, in_org_0
    in_0 = read_tree_array(f)
    in_org_0 = copy.deepcopy(in_0)
    global out
    out = read_int_matrix(f)
    f.close

def judge():
    load_test()
    capture_stdout()
    start_time = datetime.datetime.now()
    for i in range(num_test):
        print ('Testing case #' + str(i+1))
        answer = bst_to_list(in_0[i]) 
        if (not check_dlist(out[i], answer)):
            release_stdout()
            out_str = str(i+1) + " / " + str(num_test) + ";"
            out_str += str(in_org_0[i])
            out_str += ";"
            out_str += dlist_to_string(answer)
            out_str += ";"
            out_str += str(out[i])
            print(out_str)
            return

    release_stdout()
    delta = datetime.datetime.now() - start_time
    runtime = str(int(delta.total_seconds() * 1000))
    print('Accepted;' + runtime)

File: judge/ruby/tests/bst-to-dlist.rb
@num_test = 100
@in_0 = []
@in_org_0 = []
@out = []

def check_dlist(arr, list)
	cur, pre, i = list, nil, 0

	while(!cur.nil?)
		return false if i >= arr.length || arr[i] != cur.val
		i += 1
		pre = cur
		cur = cur.right
	end
	return false if i != arr.length
	cur = pre
	i -= 1
	while(!cur.nil?)
		return false if i < 0 || arr[i] != cur.val
		i -= 1
		last = cur
		cur = cur.left
		last.left = nil
	end
	return false if i != -1
	true
end

def dlist_to_string(list)
	ret = '['
	first = true
	while(!list.nil?)
		ret += ', ' if !first
		ret += list.val.to_s
		list = list.right
		first = false
	end
	ret + ']'
end

def load_test
    f = File.new("judge/tests/bst-to-dlist.txt")
    @in_0 = read_tree_array(f)
    @in_org_0 = @in_0.dup
    @out = read_int_matrix(f)
    f.close
end

def judge
    load_test
    capture_stdout

    start_time = Time.now

    (0...@num_test).each do |i|
       puts 'Testing case #' + (i+1).to_s
        answer = bst_to_list(@in_0[i]) 
        if (!check_dlist(@out[i], answer))
            release_stdout
            print "#{i+1} / #{@num_test};"
            print @in_org_0[i].to_s
            print ';'
            print dlist_to_string(answer)
            print ';'
            print @out[i].to_s
            puts
            return
        end
    end

    release_stdout
    runtime = (Time.now - start_time) * 1000.0
    puts('Accepted;' + runtime.to_i.to_s)
end
